pkg/profile/artifact_test.go

```diff
@@ -413,6 +395,156 @@ status: {}
+									Prune:           true,
+									TargetNamespace: "default",
+								},
+							},
+						},
+						Name:         "test-artifact-2",
+						RepoURL:      "https://repo-url.com",
+						PathsToCopy:  []string{"nginx2/deployment"},
+						SparseFolder: "weaveworks-nginx-2",
+						Branch:       "main",
+					},
+				}
+				pSub = profilesv1.ProfileInstallation{
+					TypeMeta: profileTypeMeta,
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      installationName,
+						Namespace: namespace,
+					},
+					Spec: profilesv1.ProfileInstallationSpec{
+						Source: &profilesv1.Source{
+							URL:    "https://github.com/weaveworks/nginx-profile",
+							Branch: "main",
+						},
+					},
+				}
+				fakeGitClient.CloneStub = func(url string, branch string, dir string) error {
+					from := filepath.Join("testdata", "clone_cache")
+					err := copy.Copy(from, filepath.Join(dir))
+					Expect(err).NotTo(HaveOccurred())
+					return nil
+				}
+			})
+
+			It("shouldn't clone as many times as there are artifacts", func() {
+				fakeGitClient.CloneStub = func(url string, branch string, dir string) error {
+					from := filepath.Join("testdata", "clone_cache")
+					err := copy.Copy(from, filepath.Join(dir))
+					Expect(err).NotTo(HaveOccurred())
+					return nil
+				}
+				profile.SetProfileMakeArtifacts(func(pam *profile.ProfilesArtifactsMaker, installation profilesv1.ProfileInstallation) ([]artifact.Artifact, error) {
+					return artifacts, nil
+				})
+				maker := profile.NewProfilesArtifactsMaker(profile.MakerConfig{
+					GitClient:        fakeGitClient,
+					RootDir:          rootDir,
+					GitRepoNamespace: gitRepoNamespace,
+					GitRepoName:      gitRepoName,
+				})
+				err := maker.Make(pSub)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(fakeGitClient.CloneCallCount()).To(Equal(1), "cache did not work, clone has been called more than once.")
+			})
+			It("should still clone different urls", func() {
+				fakeGitClient.CloneStub = func(url string, branch string, dir string) error {
+					from := filepath.Join("testdata", "clone_cache")
+					err := copy.Copy(from, filepath.Join(dir))
+					Expect(err).NotTo(HaveOccurred())
+					return nil
+				}
+				artifacts[0].RepoURL = "https://github.com/weaveworks/profiles-examples"
+				profile.SetProfileMakeArtifacts(func(pam *profile.ProfilesArtifactsMaker, installation profilesv1.ProfileInstallation) ([]artifact.Artifact, error) {
+					return artifacts, nil
+				})
+				maker := profile.NewProfilesArtifactsMaker(profile.MakerConfig{
+					GitClient:        fakeGitClient,
+					RootDir:          rootDir,
+					GitRepoNamespace: gitRepoNamespace,
+					GitRepoName:      gitRepoName,
+				})
+				err := maker.Make(pSub)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(fakeGitClient.CloneCallCount()).To(Equal(2))
+			})
+			It("should still clone same urls but with different branch", func() {
+				fakeGitClient.CloneStub = func(url string, branch string, dir string) error {
```

Comment:
can we not make this smarter, just checkout branch if repo is already cloned? or is that not really any faster?

Reply:
That would increase the complexity of this very simple thing. I traded that for the fact that I don't think it happens often enough. WDYT?

Reply:
I mean, not by that much, but I would have to also keep track of what branch is checked out, and then check out the right branch that is needed. Or always checkout the branch when I'm in the right place. It just adds a bit of more logic.

Reply:
Also, then it's no longer a shallow clone. Because I need all branches and all tags.

Reply:
Unless, I do a fetch first... then checkout. You see how it gets more complex? :D

Reply:
Also, it actually wouldn't matter in terms of performance, because we shallow clone a specific tag anyways. Which means, we have to check out / fetch that specific tag. It wouldn't be a difference since in both cases it clones the same amount -metadata I guess. Which is only a couple bytes.

Reply:
So there you have your answer. :D :D I knew I already went through this, but failed to write it down so I had to think it through again :D